kv/kvserver: TestReliableIntentCleanup failed

kv/kvserver.TestReliableIntentCleanup [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3034658&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3034658&tab=artifacts#/) on master @ [5fb0e29d1da5d09d1d869c2324d14160c0fbd71f](https://github.com/cockroachdb/cockroach/commits/5fb0e29d1da5d09d1d869c2324d14160c0fbd71f):


```
=== RUN   TestReliableIntentCleanup
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReliableIntentCleanup135291783
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestReliableIntentCleanup
    intent_resolver_integration_test.go:655: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReliableIntentCleanup135291783
--- FAIL: TestReliableIntentCleanup (529.76s)
=== RUN   TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000
        --- FAIL: TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000 (292.62s)
=== RUN   TestReliableIntentCleanup/ForceSyncIntentResolution=true
    --- FAIL: TestReliableIntentCleanup/ForceSyncIntentResolution=true (334.20s)
=== RUN   TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true/finalize=cancel/abort=push
    intent_resolver_integration_test.go:586: 
        	Error Trace:	intent_resolver_integration_test.go:379
        	            				intent_resolver_integration_test.go:586
        	            				intent_resolver_integration_test.go:643
        	            				subtest.go:34
        	Error:      	Condition never satisfied
        	Test:       	TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true/finalize=cancel/abort=push
        	Messages:   	intent cleanup timed out
    intent_resolver_integration_test.go:586: 
        	Error Trace:	intent_resolver_integration_test.go:385
        	            				intent_resolver_integration_test.go:586
        	            				intent_resolver_integration_test.go:643
        	            				subtest.go:34
        	Error:      	found stale intents
        	Test:       	TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true/finalize=cancel/abort=push
        	Messages:   	count=729 first={{"key\x00\x0e$v\x86"} id=d4c80b11 key="key\x00\x0f*\xdc#" pri=0.00075833 epo=0 ts=1622510816.699793887,0 min=1622510816.699793887,0 seq=3247} last={{"key\x00\x0f\xff\x9a^"} id=d4c80b11 key="key\x00\x0f*\xdc#" pri=0.00075833 epo=0 ts=1622510816.699793887,0 min=1622510816.699793887,0 seq=6442}
                        --- FAIL: TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true/finalize=cancel/abort=push (65.56s)
=== RUN   TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true/finalize=cancel
                    --- FAIL: TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true/finalize=cancel (147.49s)
=== RUN   TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true
                --- FAIL: TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true/txn=true (240.71s)
=== RUN   TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true
            --- FAIL: TestReliableIntentCleanup/ForceSyncIntentResolution=true/numKeys=100000/singleRange=true (243.08s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestReliableIntentCleanup PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/kv erikgrinaker
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReliableIntentCleanup.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Look at how long this ran:

--- FAIL: TestReliableIntentCleanup (529.76s)

I'd wager a guess you just haven't pushed it into that territory? Also, some tests flake much more frequently in CI because in CI packages are tested in parallel, whereas locally we usually stress "just that one test". However stressing the package is very annoying, so I'd hope that playing with `STRESSFLAGS='-p 100'` on roachprod-stress or something like that will do it?

--- a/pkg/kv/kvserver/intent_resolver_integration_test.go
+++ b/pkg/kv/kvserver/intent_resolver_integration_test.go
@@ -15,6 +15,7 @@ import (
 	"context"
 	"encoding/binary"
 	"fmt"
+	"math"
 	"math/rand"
 	"sync"
 	"testing"
@@ -33,6 +34,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/assert"
@@ -219,7 +221,6 @@ func TestRollbackSyncRangedIntentResolution(t *testing.T) {
 func TestReliableIntentCleanup(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	skip.WithIssue(t, 65908, "flaky in CI")
 	skip.UnderShort(t) // takes 294s
 	skip.UnderRace(t, "timing-sensitive test")
 	skip.UnderStress(t, "memory-hungry test")
@@ -372,24 +373,50 @@ func TestReliableIntentCleanup(t *testing.T) {
 			}
 		}
 
-		// assertIntentCleanup checks that intents get cleaned up within a
-		// reasonable time.
+		// assertIntentCleanup checks that intents get cleaned up. It errors
+		// ~fast if the intent count does not decrease, but gives it ample time
+		// to complete when progress is being made.
 		assertIntentCleanup := func(t *testing.T) {
 			t.Helper()
-			var result storage.MVCCScanResult
-			if !assert.Eventually(t, func() bool {
-				result, err = storage.MVCCScan(ctx, store.Engine(), prefix, prefix.PrefixEnd(),
+			const (
+				checkInterval  = time.Second
+				stuckTimeout   = 30 * time.Second // intent count not decreasing
+				overallTimeout = 5 * time.Minute
+			)
+			var (
+				lastIntentDecrease = timeutil.Now()
+				lastIntentCount    = math.MaxInt64
+				started            = timeutil.Now()
+			)
+			for {
+				result, err := storage.MVCCScan(ctx, store.Engine(), prefix, prefix.PrefixEnd(),
 					hlc.MaxTimestamp, storage.MVCCScanOptions{Inconsistent: true})
 				require.NoError(t, err)
-				return len(result.Intents) == 0
-			}, time.Minute, 200*time.Millisecond, "intent cleanup timed out") {
-				require.Fail(t, "found stale intents", "count=%v first=%v last=%v",
-					len(result.Intents), result.Intents[0], result.Intents[len(result.Intents)-1])
+				intentCount := len(result.Intents)
+				if intentCount == 0 {
+					return
+				}
+				if intentCount < lastIntentCount {
+					lastIntentDecrease = timeutil.Now()
+				}
+				lastIntentCount = intentCount
+				if timeutil.Since(lastIntentDecrease) >= stuckTimeout {
+					require.Fail(t, "found stale intents", "count=%v first=%v last=%v",
+						len(result.Intents), result.Intents[0], result.Intents[len(result.Intents)-1])
+				}
+				if timeutil.Since(started) >= overallTimeout {
+					require.Fail(t, "intent cleanup timed out", "count=%v first=%v last=%v",
+						len(result.Intents), result.Intents[0], result.Intents[len(result.Intents)-1])
+				}
+				time.Sleep(checkInterval)
 			}
 		}
 
 		// assertTxnCleanup checks that the txn record is cleaned up within a
-		// reasonable time.
+		// reasonable time. We give it a long timeout, since there are cases
+		// where it will attempt to clean up all the intents again (even though
+		// they have already been removed), which takes time to process. This
+		// can happen due to multiple rollbacks being sent by different actors.
 		assertTxnCleanup := func(t *testing.T, txnKey roachpb.Key, txnID uuid.UUID) {
 			t.Helper()
 			var txnEntry roachpb.Transaction
@@ -399,7 +426,7 @@ func TestReliableIntentCleanup(t *testing.T) {
 					storage.MVCCGetOptions{})
 				require.NoError(t, err)
 				return !ok
-			}, 10*time.Second, 100*time.Millisecond, "txn record cleanup timed out") {
+			}, time.Minute, 100*time.Millisecond, "txn record cleanup timed out") {
 				require.Fail(t, "found stale txn record", "%v", txnEntry)
 			}
 		}
@@ -586,13 +613,6 @@ func TestReliableIntentCleanup(t *testing.T) {
 
 			assertIntentCleanup(t)
 			for txnID, txnKey := range txns {
-				// TODO(erikgrinaker): this occasionally fails in CI with
-				// finalize=cancelAsync, but the reason hasn't been found. We're
-				// mostly concerned with intent cleanup, so skip the txn record
-				// assertion in these cases.
-				if spec.finalize == "cancelAsync" {
-					break
-				}
 				assertTxnCleanup(t, txnKey, txnID)
 			}
 		}

